### ISSUE TITLE
Implement the correct rules for whether overridden methods should be documented

### ISSUE THREAD

**Issue by timhunt** [moodlehq/moodle-local_moodlecheck#125]
Moodle coding style says that this is how it should be:

```php
/** ... */
class base_class {
    /**
     * Document the purpose and contract of this method.
     */
    funcition do_something(): void {
    }
}
```

```php
/** ... */
class sub_class extends base_class {
    funcition do_something(): void {
        // Just implement the purpopse defined in the base class, so there should not be a duplicate PHPdoc comment here.
    }
}
```

However, the PHPdoc checker currently gets this wrong, becuase it only analyses one file at a time.

However, now we rely more on class autoloading, we might finally be able to improve this, at least in the case where:

1. this file contains a single class definition.
2. This class extends classes/implements interfaces which are autoloadable.

Then, we can parse those files to get a list of public and protected methods, and we can use that in local_moodlecheck_functionsdocumented, to avoid giving warnigns about those funcitions. (Indeed, we could start issuing a warning if there is a comment there.)

**Comment by james-cnz** [comment]
I think there was an intention to reimplement these checks in PHPCS at some stage?  And AFAIK PHPCS only scans one file at a time?  So I'm not sure this would be possible.

**Comment by timhunt** [comment]
You are right that becuase PHPCS only scans one file at a time, this is effectively impossible.

However, a practical solution was found, becuase PHP introduced the `#[\Override]` annotation. So, now in Moodle, overridden methods should have that annotation, and if a method has that, CodeChecker will not want a PHPdoc comment. Correct use of `#[\Override]` is checked by most IDEs.

So, I will close this now.
